### PR TITLE
Sort colors

### PR DESCRIPTION
**Pull request by Mohamed-Elnageeb**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

**Review by linahKhayri — COMMENTED** [review]
Hey Mohamed, reviewing your code was a good experience, at first I thought the problem might be something new to me, but the comments at each step and the docstring explaining the behavior of your code helped me understand everything and it turned out simple!. It’s well-organized and structured, and it passed all the checklists
great job! 👌
I just added a few comments in some lines of the code that I thought would make it clearer, but over all it's AMAZING👏

File: solutions/tests/test_sort_colors.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        with self.assertRaises(AssertionError):
+            sort_colors("not a list")
+
+    def test_invalid_elements(self):
+        """It should raise an assertion error for invalid elements"""
+        nums = [0, 1, 3]  # Invalid element 3
+        with self.assertRaises(AssertionError):
+            sort_colors(nums)
+
+    def test_randomly_generated_list(self):
```
[the remark this answers is not in the document]
Comment (Mohamed-Elnageeb) [reply solutions/tests/test_sort_colors.py:90]: Ok thanks....also I think I will remove this test as i realised this can fall under "using logic in a unit test".

File: solutions/sort_colors.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    >>> sort_colors(nums)
+    >>> nums
+    [0, 1, 1, 2, 2]
+    """
+    # Validate input
+    assert isinstance(nums, list), "Input must be a list."
+    assert all(0 <= number <= 2 for number in nums), "All elements must be 0, 1, or 2."
+
+    # Create a counter array to store the count of each number
+    counter_array = [0] * 3
```
[the remark this answers is not in the document]
Comment (Mohamed-Elnageeb) [reply solutions/sort_colors.py:69]: Ok,Will do tomorrow 😇

**Review by linahKhayri — DISMISSED** [review]
Good job Mohamed👏

File: solutions/sort_colors.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    assert all(0 <= number <= 2 for number in nums), "All elements must be 0, 1, or 2."
+
+    # Create a counter array to store the count of each number
+    counter_list = [0] * 3
+
+    # Count occurrences of each number
+    for number in nums:
+        counter_list[number] += 1
+
+    # Rebuild the nums array in-place
```
Comment (GaiSamuel) [inline solutions/sort_colors.py:75]: This too, "array" to "list"

**Review by GaiSamuel — CHANGES_REQUESTED** [review]
Well, elaborated! I have suggested just a few changes and enquired an explanation on line 91-92

File: solutions/tests/test_sort_colors.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+            sort_colors("not a list")
+
+    def test_invalid_elements(self):
+        """It should raise an assertion error for invalid elements"""
+        nums = [0, 1, 3]  # Invalid element 3
+        with self.assertRaises(AssertionError):
+            sort_colors(nums)
+
+
+if __name__ == "__main__":
```
[the remark this answers is not in the document]
Comment (Mohamed-Elnageeb) [reply solutions/tests/test_sort_colors.py:91]: This statement "if __name__ == "__main__"  is used when you want code executed only when a file is run as a script rather than imported as a module. 
So i want the tests to run.and btw when i tried witgout it i got py_tests errors during CI checks.

File: solutions/sort_colors.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    >>> nums = [2, 1, 1, 0, 2]
+    >>> sort_colors(nums)
+    >>> nums
+    [0, 1, 1, 2, 2]
+    """
+    # Validate input
+    assert isinstance(nums, list), "Input must be a list."
+    assert all(0 <= number <= 2 for number in nums), "All elements must be 0, 1, or 2."
+
+    # Create a counter array to store the count of each number
```
[the remark this answers is not in the document]
Comment (Mohamed-Elnageeb) [reply solutions/sort_colors.py:68]: Ok

File: solutions/tests/test_sort_colors.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (GaiSamuel) [reply solutions/tests/test_sort_colors.py:91]: oh alright.

File: solutions/tests/test_sort_colors.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+import random
+import unittest
+
+# Import the function to be tested
+from ..sort_colors import sort_colors
+
+
+class TestSortColors(unittest.TestCase):
+    """Test the sort_colors function"""
+
```
[the remark this answers is not in the document]
Comment (Khusro-S) [reply solutions/tests/test_sort_colors.py:41]: Yea I agree, would make it more easier to differentiate between the cases but at the same time, might be redundant 🤔

File: solutions/tests/test_sort_colors.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        nums = [2, 0, 2, 1, 1, 0]
+        sort_colors(nums)
+        self.assertEqual(nums, [0, 0, 1, 1, 2, 2])
+
+    def test_large_list(self):
+        """It should handle a large list"""
+        nums = [0, 2, 1] * 100000
+        expected = sorted(nums)  # expected output is a sorted array
+        sort_colors(nums)
+        self.assertEqual(nums, expected)
```
Comment (Khusro-S) [inline solutions/tests/test_sort_colors.py:77]: Even though it is okay but maybe try hardcoding the values, but then it'd probably be hundreds or thousands of lines of code 🤔.... yea this might be better in this situation nvm😄

**Review by Khusro-S — APPROVED** [review]
This function basically sorts a list, which is only supposed to contain numbers `0`, `1`, and `2`. The counter_list counts how many `0`s, `1`s, and `2`s are in the list. Then, it places those numbers back into the original list in sorted order.

For example, if our input list is:
`[1, 1, 2, 1, 0, 0, 1]`

The counter_list will count the occurrences of each number:
`counter_list = [2, 4, 1]`

This means there are two `0`s, four `1`s, and one `2`  in the list.
Next, the rebuilding part places the numbers back into the original list. Here's how it works step by step:

1. Start by adding all the `0`s → `[0, 0, 2, 1, 0, 0, 1]`
2. Then, add all the `1`s → `[0, 0, 1, 1, 1, 1, 1]`
3. Finally, add the `2`s → `[0, 0, 1, 1, 1, 1, 2]`
The final sorted list is:
`[0, 0, 1, 1, 1, 1, 2]`

This explanation is just to help me understand the implementation and confirm whether I’ve understood it correctly. Please let me know if I missed anything or got something wrong.

 I really liked how this implementation avoids using built-in sorting functions and sticks to an efficient and straightforward counting sort approach. The logic is super clear, and the in-place sorting is a nice touch! Great job! 🎉

**Review by GaiSamuel — APPROVED** [review]
Looks good. Great work.
